pkg/frontend/querymiddleware/astmapper/pruning_new_test.go

```diff
@@ -0,0 +1,179 @@
+			`vector(0)`,
+		},
+		{
+			`up < -Inf`,
+			`up < -Inf`,
+		},
+		{
+			`avg(rate(foo[1m])) < (-Inf)`,
+			`avg(rate(foo[1m])) < (-Inf)`,
+		},
+		{
+			`Inf * -1`,
+			`Inf * -1`,
+		},
+		{
+			`avg(rate(foo[1m])) < (-1 * +Inf)`,
+			`avg(rate(foo[1m])) < (-1 * +Inf)`,
+		},
+		{
+			`(-1 * -Inf) < avg(rate(foo[1m]))`,
+			`(-1 * -Inf) < avg(rate(foo[1m]))`,
+		},
+		{
+			`vector(0) < -Inf or avg(rate(foo[1m]))`,
+			`vector(0) < -Inf or avg(rate(foo[1m]))`,
+		},
+		{
+			`avg(rate(foo[1m])) or vector(0) < -Inf`,
+			`avg(rate(foo[1m])) or vector(0) < -Inf`,
+		},
+		{
+			`avg(rate(foo[1m])) or (vector(0) < -Inf)`,
+			`avg(rate(foo[1m])) or (vector(0) < -Inf)`,
+		},
+		{
+			`((-1 * -Inf) < avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`((-1 * -Inf) < avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`(avg(rate(foo[2m])) < (+1 * -Inf)) or avg(rate(foo[1m]))`,
+			`(avg(rate(foo[2m])) < (+1 * -Inf)) or avg(rate(foo[1m]))`,
+		},
+		{
+			`((-1 * -Inf) == avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`((-1 * -Inf) == avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`avg(rate(foo[1m])) unless ((-1 * -Inf) < avg(rate(foo[2m])))`,
+			`avg(rate(foo[1m])) unless ((-1 * -Inf) < avg(rate(foo[2m])))`,
+		},
+		{
+			`((2 * +Inf) < avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`((2 * +Inf) < avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`(((-1 * -Inf) < avg(rate(foo[3m]))) unless avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`(((-1 * -Inf) < avg(rate(foo[3m]))) unless avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`((((-1 * -Inf) < avg(rate(foo[4m]))) unless avg(rate(foo[3m]))) and avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+			`((((-1 * -Inf) < avg(rate(foo[4m]))) unless avg(rate(foo[3m]))) and avg(rate(foo[2m]))) or avg(rate(foo[1m]))`,
+		},
+		{
+			`avg(rate(foo[1m])) or avg(rate(bar[1m]))`,
+			`avg(rate(foo[1m])) or avg(rate(bar[1m]))`,
+		},
+		{ // The const expression is on the wrong side.
+			`(vector(0) == 1) and on() (avg(rate(foo[1m])))`,
+			`(vector(0) == 1) and on() (avg(rate(foo[1m])))`,
+		},
+		{ // Matching on labels.
+			`(avg(rate(foo[1m]))) and on(id) (vector(0) == 1)`,
+			`(avg(rate(foo[1m]))) and on(id) (vector(0) == 1)`,
+		},
+		{ // Not "on" expression.
+			`(avg(rate(foo[1m]))) and ignoring() (vector(0) == 1)`,
+			`(avg(rate(foo[1m]))) and ignoring() (vector(0) == 1)`,
+		},
+		// Pruned expressions.
+		{
+			`(avg(rate(foo[1m]))) and on() (vector(0) == 1)`,
+			`(vector(0) == 1)`,
+		},
+		{
+			`(avg(rate(foo[1m]))) and on() (vector(1) == 1)`,
+			`(avg(rate(foo[1m])))`,
+		},
+		{
+			`(avg(rate(foo[1m]))) and on() (vector(3) == 4.5)`,
+			`(vector(3) == 4.5)`,
+		},
+		{
+			`(avg(rate(foo[1m]))) and on() (vector(5.5) == 5.5)`,
+			`(avg(rate(foo[1m])))`,
+		},
+		{
+			// "and on()" is not on top level, "or" has lower precedence.
+			`(avg(rate(foo[1m]))) and on() (vector(0) == 1) or avg(rate(bar[1m]))`,
+			`(vector(0) == 1) or avg(rate(bar[1m]))`,
```

Comment:
Couldn't we simplify this further, to `avg(rate(bar[1m]))`?

Reply:
That's more complicated. The main point is to avoid loading chunks and `(vector(0)==1)` doesn't load chunks, so should be very  efficient.
We can add that logic in a new PR, seems to be independent from this optimization, wdyt?
But having two algorithms raises the question of whether they need to be run repeatedly to optimize away everything, so it gets even more complicated.

Reply:
Note that removing `(vector(0)==1)` from an OR expression is different from removing it from an AND expression, which is what I mean by more complicated . Outside my scope.

Reply:
Is there a test for this? We should document that in a test so nobody comes tomorrow and "optimizes further".

Reply:
>having two algorithms raises the question of whether they need to be run repeatedly to optimize away everything, so it gets even more complicated.

If we optimise the leaf node expressions first and work our way back up the tree, can't we do both `and` and `or` in one pass?

But if you want to limit this to `and` and come back to `or` later, that's fine by me, please just make this clear with a test or a comment as Oleg suggests.

Reply:
adding both

Reply:
well, actually I had 3 tests on this already so just improved the comments